src/CommunityToolkit.Maui/PopupService.cs

```diff
@@ -126,24 +136,52 @@ public PopupService()
 	/// <exception cref="InvalidOperationException"></exception>
 	static void ValidateBindingContext<TViewModel>(Popup popup, out TViewModel bindingContext)
 	{
-		if (popup.BindingContext is not TViewModel viewModel)
+		if (popup.BindingContext is TViewModel viewModel)
+		{
+			bindingContext = viewModel;
+			return;			
+		}
+		else if (popup.Content?.BindingContext is TViewModel contentViewModel)
 		{
-			throw new InvalidOperationException($"Unexpected type has been assigned to the BindingContext of {popup.GetType().FullName}. Expected type {typeof(TViewModel).FullName} but was {popup.BindingContext?.GetType().FullName ?? "null"}");
+			bindingContext = contentViewModel;
+			return;
 		}
 
-		bindingContext = viewModel;
+		throw new InvalidOperationException($"Unexpected type has been assigned to the BindingContext of {popup.GetType().FullName}. Expected type {typeof(TViewModel).FullName} but was {popup.BindingContext?.GetType().FullName ?? "null"}");
 	}
 
 	Popup GetPopup(Type viewModelType)
 	{
-		var popup = serviceProvider.GetService(viewModelToViewMappings[viewModelType]) as Popup;
+		var view = serviceProvider.GetService(viewModelToViewMappings[viewModelType]);
 
-		if (popup is null)
+		if (view is null)
 		{
 			throw new InvalidOperationException(
 				$"Unable to resolve popup type for {viewModelType} please make sure that you have called {nameof(AddTransientPopup)}");
 		}
 
+		Popup popup;
+
+		if (view is View visualElement)
+		{
+			popup = new Popup
+			{
+				Content = visualElement
+			};
+
+			// Binding the BindingContext property up from the view to the popup so that it is nicely handled on macOS.
+			popup.SetBinding(Popup.BindingContextProperty, new Binding { Source = view, Path = Popup.BindingContextProperty.PropertyName });
+		}
+		else if (view is Popup viewPopup)
```

Comment:
let's make a breaking change with this implementation and make Popup an internal class. it should simplify popup service implementation

Reply:
This might be the end goal but if we are going to make this breaking change I think we should mark Popup as `[Obsolete]` in the next release to give developers a heads up.

Reply:
I would say that we can't do that... Look, the Popup is the control that is used by handlers, and because of that, they are a point of extension if anyone wants to customize the Popup control (adding properties, overriding behaviors, etc), so making this control `internal` will be against .NET MAUI patterns, holding users into a control that they can't modify.

Even with this addition, I would say to keep this control public.

Reply:
Let's leave it as is for now and see how people adopt this approach